platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,437 @@
+# TSS - Roster Proposal
+
+---
+
+## Summary
+
+This proposal outlines the design specification of rosters to support Threshold Signature Scheme (TSS) block signing.
+It details the roster's lifecycle, data storage, API, and necessary changes to associated components within the
+platform.
+
+| Metadata           | Entities                                      | 
+|--------------------|-----------------------------------------------|
+| Designers          | Kore, Anthony, Ed, Cody, Austin               |
+| Functional Impacts | TSS, Genesis, Cryptography, Reconnect         |
+| Related Proposals  | TSS-Library, TSS-Ledger-Id, TSS-Block-Signing |
+| HIPS               | N/A                                           |
+
+## Purpose and Context
+
+The introduction of the Threshold Signature Scheme (TSS) requires a new mechanism for managing node participation in
+consensus and block signing.
+The Roster, an immutable subset of the address book, will provide this mechanism, ensuring efficient and secure key
+management for TSS operations.
+This proposal provides a specification for the behavior of rosters starting from their creation from the
+Candidate Address Book (CAB) to their terminal state within the TSS specification.
+A roster reaches a terminal state when it is either adopted by the platform or replaced by a new roster.
+
+The Future Address Book (FAB) is a dynamic list maintained within the App that reflects the desired future state of the
+network's nodes. It will be continuously updated by HAPI transactions, such as those that create, update, or delete
+nodes.
+The CAB is a snapshot of the Future Address Book (FAB), and it is created when the app decides to adopt a new address
+book.
+
+This proposal specifies that a Roster will be created from the CAB by the Hedera app and passed to the
+platform code.
+The mechanism for doing so is detailed below.
+
+### Requirements
+
+- Immutability: Rosters must be immutable to protect the integrity of its computed hash.
+- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
+  copying.
+- Clear API: A well-defined API should be provided for the app to create and submit candidate rosters, and for the
+  platform to manage the lifecycle of submitted rosters.
+- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
+  Event Validation etc.) must be adapted to use rosters instead.
+    -
+- Address book paradigm discontinued within the platform codebase, only use rosters instead
+
+### Architecture and/or Components
+
+###### Roster Lifecycle
+
+![](TSS%20Roster%20Lifecycle.drawio.svg)
+
+###### Roster Public API
+
+A new method will be added to the platform API to allow the App to submit a Candidate Roster to the platform:
+
+```java
+//in SwirldState
+
+void setCandidateRoster(@NonNull final Roster candidateRoster);
```

Comment:
We should document if there are any limitations regarding the thread on which this method can be called (if there are any).

Reply:
From Platform code's perspective, setting a candidate roster, validating and storing it is expected to be single threaded, with that thread being owned by the Services code that initiates the operation. If i get your question right, I'm not aware of any limitation, maybe other than the single threadedness of it - rosters can be set only one at a time.